recipe/multitenancy/recipe.go

```diff
@@ -0,0 +1,136 @@
+	r.Config = verifiedConfig
+
+	querierInstance, err := supertokens.GetNewQuerierInstanceOrThrowError(recipeId)
+	if err != nil {
+		return nil, err
+	}
+	recipeImplementation := makeRecipeImplementation(*querierInstance, verifiedConfig, appInfo)
+	r.RecipeImpl = verifiedConfig.Override.Functions(recipeImplementation)
+
+	r.APIImpl = verifiedConfig.Override.APIs(api.MakeAPIImplementation())
+
+	recipeModuleInstance := supertokens.MakeRecipeModule(recipeId, appInfo, r.handleAPIRequest, r.getAllCORSHeaders, r.getAPIsHandled, nil, r.handleError, onSuperTokensAPIError)
+	r.RecipeModule = recipeModuleInstance
+
+	r.staticThirdPartyProviders = []tpmodels.ProviderInput{}
+
+	return r, nil
+}
+
+func GetRecipeInstanceOrThrowError() (*Recipe, error) {
+	if singletonInstance != nil {
+		return singletonInstance, nil
+	}
+
+	return nil, errors.New("Initialisation not done. Did you forget to call the init function?")
+}
+
+func recipeInit(config *multitenancymodels.TypeInput) supertokens.Recipe {
+	return func(appInfo supertokens.NormalisedAppinfo, onSuperTokensAPIError func(err error, req *http.Request, res http.ResponseWriter)) (*supertokens.RecipeModule, error) {
+		if singletonInstance == nil {
+			recipe, err := MakeRecipe(RECIPE_ID, appInfo, config, onSuperTokensAPIError)
+			if err != nil {
+				return nil, err
+			}
+			singletonInstance = recipe
+			return &singletonInstance.RecipeModule, nil
+		}
+		return nil, errors.New("Multitenancy recipe has already been initialised. Please check your code for bugs.")
+	}
+}
+
+// implement RecipeModule
+
+func (r *Recipe) getAPIsHandled() ([]supertokens.APIHandled, error) {
+	loginMethodsAPI, err := supertokens.NewNormalisedURLPath(LoginMethodsAPI)
+	if err != nil {
+		return nil, err
+	}
+
+	return []supertokens.APIHandled{
+		{
+			Method:                 http.MethodGet,
+			PathWithoutAPIBasePath: loginMethodsAPI,
+			ID:                     LoginMethodsAPI,
+			Disabled:               r.APIImpl.LoginMethodsGET == nil,
+		},
+	}, nil
+}
+
+func (r *Recipe) handleAPIRequest(id string, req *http.Request, res http.ResponseWriter, theirHandler http.HandlerFunc, _ supertokens.NormalisedURLPath, _ string) error {
+	options := multitenancymodels.APIOptions{
+		RecipeImplementation:      r.RecipeImpl,
+		Config:                    r.Config,
+		RecipeID:                  RECIPE_ID,
+		Req:                       req,
+		Res:                       res,
+		OtherHandler:              theirHandler,
+		StaticThirdPartyProviders: r.staticThirdPartyProviders,
+	}
+	if id == LoginMethodsAPI {
+		return api.LoginMethodsAPI(r.APIImpl, options)
+	}
+	return errors.New("should never come here")
+}
+
+func (r *Recipe) getAllCORSHeaders() []string {
+	return []string{}
+}
+
+func (r *Recipe) handleError(err error, req *http.Request, res http.ResponseWriter) (bool, error) {
+	return false, nil
+}
+
+func ResetForTest() {
+	singletonInstance = nil
+}
+
+func init() {
+	supertokens.DefaultMultitenancyRecipe = recipeInit(nil)
+}
```

Comment:
add comments explaining why this is there

Reply:
Done